annotation/tests/test_annotations.py

```diff
@@ -235,3 +274,188 @@ def test_create_manifest_json(
             annotated_doc.tenant,
             moto_s3,
         )
+
+
+def test_construct_annotated_doc(
+    annotated_doc: AnnotatedDoc,
+    annotation_doc_for_save: DocForSaveSchema,
+):
+    doc = annotation_doc_for_save
+    latest_doc = annotated_doc
+    is_latest = False
+    new_equal_latest = False
+    s3_path = (
+        f"{S3_START_PATH}/{annotated_doc.job_id}/" f"{annotated_doc.file_id}"
+    )
+    s3_file_path = "path"
+    s3_file_bucket = "bucket"
+
+    pages_sha_non_latest = ({"1": "a"}, "c")
+
+    db = Mock()
+    db.add = Mock()
+    db.add_all = Mock()
+    db.commit = Mock()
+
+    with patch(
+        "annotation.annotations.main.get_pages_sha",
+        return_value=pages_sha_non_latest,
+    ) as mock_get_pages_sha, patch(
+        "annotation.annotations.main.check_docs_identity",
+        return_value=new_equal_latest,
+    ) as mock_check_docs, patch(
+        "annotation.annotations.main.construct_document_links", return_value=[]
+    ) as mock_construct_doc_links, patch(
+        "annotation.annotations.main.convert_bucket_name_if_s3prefix",
+        return_value=TEST_TENANT,
+    ) as mock_convert_bucket, patch(
+        "annotation.annotations.main.upload_pages_to_minio"
+    ) as mock_upload_pages, patch(
+        "annotation.annotations.main.create_manifest_json"
+    ) as mock_create_manifest:
+        construct_annotated_doc(
+            db,
+            annotated_doc.user,
+            None,
+            annotated_doc.job_id,
+            annotated_doc.file_id,
+            doc,
+            TEST_TENANT,
+            s3_file_path,
+            s3_file_bucket,
+            latest_doc,
+            1,
+            is_latest,
+        )
+
+        mock_get_pages_sha.assert_called_once_with(
+            doc.pages,
+            latest_doc.revision,
+            doc.validated,
+            doc.failed_validation_pages,
+            doc.user,
+        )
+        mock_check_docs.assert_called_once()
+
+        mock_construct_doc_links.assert_called_once()
+        db.add.assert_called_once()
+        db.add_all.assert_called_once_with([])
+        db.commit.assert_called_once()
+        mock_convert_bucket.assert_called_once_with(TEST_TENANT)
+        mock_upload_pages.assert_called_once_with(
+            pages=doc.pages,
+            pages_sha=pages_sha_non_latest[0],
+            s3_path=s3_path,
+            bucket_name=TEST_TENANT,
+            s3_resource=None,
+        )
+        mock_create_manifest.assert_called_once_with(
+            ANY,
```

Comment:
Here we encounter the same problem as with the asserting the actual output of the function. Since assert compares object **identity** and not object attributes. I would need to define a `__eq__ `method inside of `AnnotatedDoc `model for it to be able to work.  Do we have any better ideas than that?